File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/CreateGame_Notification.java
package org.emulinker.kaillera.controller.v086.protocol;

import com.google.auto.value.AutoValue;
import org.emulinker.kaillera.controller.messaging.MessageFormatException;

@AutoValue
public abstract class CreateGame_Notification extends CreateGame {
  private static final String DESC = "Create Game Notification";

  public static AutoValue_CreateGame_Notification create(
      int messageNumber, String username, String romName, String clientType, int gameId, int val1)
      throws MessageFormatException {
    V086Message.validateMessageNumber(messageNumber, DESC);

    if (romName.length() == 0) {
      throw new MessageFormatException("Invalid " + DESC + " format: romName.length == 0");
    }

    if (gameId < 0 || gameId > 0xFFFF) {
      throw new MessageFormatException(
          "Invalid " + DESC + " format: gameID out of acceptable range: " + gameId);
    }

    if (val1 != 0x0000 && val1 != 0xFFFF) {
      throw new MessageFormatException(
          "Invalid " + DESC + " format: val1 out of acceptable range: " + val1);
    }

    return new AutoValue_CreateGame_Notification(
        messageNumber, ID, DESC, username, romName, clientType, gameId, val1);
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/CreateGame_Notification.kt
package org.emulinker.kaillera.controller.v086.protocol

import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.v086.protocol.V086Message.Companion.validateMessageNumber

data class CreateGame_Notification
    @Throws(MessageFormatException::class)
    constructor(
        override val messageNumber: Int,
        override val username: String,
        override val romName: String,
        override val clientType: String,
        override val gameId: Int,
        override val val1: Int
    ) : CreateGame() {

  override val messageId = ID

  init {
    validateMessageNumber(messageNumber)
    require(romName.isNotBlank()) { "romName cannot be blank" }
    require(gameId in 0..0xFFFF) { "gameID out of acceptable range: $gameId" }
    require(val1 in 0..0xFFFF) { "val1 out of acceptable range: $val1" }
  }
}
